feat(metadata) Save basic metadata to EPUB file - Bonus use Calibre Custom Columns for importing metadata.

jadehawk opened the pull request:
# 🚀 Pull Request

## 📝 Description
Currently any EPUBS added to Booklore that were processed by Calibre will have the following Metadata  extracted and placed in the appropriate Booklore's Metadata Fields:

- Title
- Language
- Authors
- Publisher
- Publish Date
- Genres
- Series Name
- Series #
- ISBN
- Comment/Description

What fails to be imported from Calibre are the following:

### Identifiers:

- asin / amazon / mobi-asin
- goodreads
- google
- hardcover
- comicvine

Besides the basic metadata, everything else only lives in Booklore's Database, In the event your DB get corrupted or we migrate to another DB, or you simply want to start from scratch, none of the rich metadata will be preserved.

<!-- Reference related issues using "Fixes #123", "Closes #456", or "Relates to #789" -->
Fixes #1878

## 🛠️ Changes Implemented
With the implemented fix on this PR, we can now populate ALL metadata from Calibre (Polished) EPUBS and we also can save the extra (ALL) metadata we can see on Booklore's UI into the EPUB file, and Booklore will be able to extract the metadata.

The Following custom columns with their assigned "lookup" names" will be detected and placed in the proper Booklore's metadata fields:

<table>
  <tr>
    <th></th>
    <th>Calibre Custom Column (#lookupname)</th>
    <th>Booklore's Metadata Field</th>
  </tr>

  <tr>
    <td></td>
    <td>Subtitle (<strong>#subtitle</strong>)</td>
    <td>Subtitle</td>
  </tr>

  <tr>
    <td></td>
    <td>Moods (<strong>#moods</strong>)</td>
    <td>Moods</td>
  </tr>

  <tr>
    <td></td>
    <td>Tags (<strong>#extra_tags</strong>)</td>
    <td>Tags</td>
  </tr>

  <tr>
    <td></td>
    <td>Series Total (<strong>#series_total</strong>)</td>
    <td>Series Total</td>
  </tr>

  <tr>
    <td></td>
    <td>Pages (<strong>#pagecount</strong>)</td>
    <td>Pages</td>
  </tr>

  <tr>
    <td></td>
    <td>Amazon Rating (<strong>#amazon_rating</strong>)</td>
    <td>Amazon Rating</td>
  </tr>

  <tr>
    <td></td>
    <td>Amazon Reviews (<strong>#amazon_review_count</strong>)</td>
    <td>Amazon Reviews</td>
  </tr>

  <tr>
    <td></td>
    <td>Goodreads Rating (<strong>#goodreads_rating</strong>)</td>
    <td>Goodreads Rating</td>
  </tr>

  <tr>
    <td></td>
    <td>Goodreads Reviews (<strong>#goodreads_review_count</strong>)</td>
    <td>Goodreads Reviews</td>
  </tr>

  <tr>
    <td></td>
    <td>Hardcover Rating (<strong>#hardcover_rating</strong>)</td>
    <td>Hardcover Rating</td>
  </tr>

  <tr>
    <td></td>
    <td>Hardcover Reviews (<strong>#hardcover_review_count</strong>)</td>
    <td>Hardcover Reviews</td>
  </tr>
</table>

<br>

<table>
  <tr>
    <th></th>
    <th>Identifier</th>
    <th>In Calibre Identifier Section</th>
    <th>Booklore's Metadata Field</th>
  </tr>
  <tr>
    <td>1</td>
    <td>ISBN</td>
    <td>isbn:XXXXXXXX</td>
    <td>ISBN-10 or ISBN-13</td>
  </tr>
  <tr>
    <td></td>
    <td>ASIN</td>
    <td>asin:XXXXXXXX</td>
    <td>Amazon ASIN</td>
  </tr>
  <tr>
    <td></td>
    <td>AMAZON</td>
    <td>amazon:XXXXXXXX</td>
    <td>Amazon ASIN</td>
  </tr>
  <tr>
    <td></td>
    <td>MOBI-ASIN</td>
    <td>mobi-asin:XXXXXXXX</td>
    <td>Amazon ASIN</td>
  </tr>
  <tr>
    <td></td>
    <td>GOODREADS</td>
    <td>goodreads:XXXXXXXX</td>
    <td>Goodreads ID</td>
  </tr>
  <tr>
    <td></td>
    <td>GOOGLE</td>
    <td>google:XXXXXXXX</td>
    <td>Google Books ID</td>
  </tr>
  <tr>
    <td></td>
    <td>HARDCOVER</td>
    <td>hardcover:XXXXXXXX</td>
    <td>Hardcover ID</td>
  </tr>
  <tr>
    <td></td>
    <td>HARDCOVER_BOOK</td>
    <td>hardcover_book:XXXXXXXX</td>
    <td>Hardcover Book ID</td>
  </tr>
  <tr>
    <td></td>
    <td>COMICVINE</td>
    <td>comicvine:XXXXXXXX</td>
    <td>Comicvine ID</td>
  </tr>
  <tr>
    <td></td>
    <td>LUBIMYCZYTAC</td>
    <td>Lubimyczytac:XXXXXXXX</td>
    <td>LC ID</td>
  </tr>
  <tr>
    <td></td>
    <td>RANOBEDB</td>
    <td>ranobedb:XXXXXXXX</td>
    <td>Ranobedb ID</td>
  </tr>
</table>
<br>

**<sup>1</sup> Special Case:**
Calibre only accepts "1" isbn, Booklore will process this `isbn` and based on character count, will place it in ISBN-10 or ISBN-13.

## Enhancements

While working on this PR, I realized that most of Booklore's Rich Metadata (Fetched from metadata providers) lives only on the database, in the event user wants to start from scratch or database gets corrupted, user will need to fetch all the metadata again.
To fix this, we will now embed all Booklore's metadata inside the EPUB, and next time you IMPORT the EPUB, Booklore will 
recognize the embedded metadata and populate the proper fields.

The following additional metadata will be saved to the EPUB file together with the standard EPUB 3 fields, and Identifiers:

<table>
  <tr>
    <th></th>
    <th>Booklore Metadata</th>
    <th>Tag</th>
  </tr>
  <tr>
    <td></td>
    <td>Moods</td>
    <td>booklore:moods</td>
  </tr>
  <tr>
    <td></td>
    <td>Tags</td>
    <td>booklore:tags</td>
  </tr>
  <tr>
    <td></td>
    <td>Series Total</td>
    <td>booklore:series_total</td>
  </tr>
  <tr>
    <td></td>
    <td>Pages</td>
    <td>booklore:page_count</td>
  </tr>
  <tr>
    <td></td>
    <td>Amazon Rating</td>
    <td>booklore:amazon_rating</td>
  </tr>
  <tr>
    <td></td>
    <td>Amazon Reviews</td>
    <td>booklore:amazon_review_count</td>
  </tr>
  <tr>
    <td></td>
    <td>Goodreads Rating</td>
    <td>booklore:goodreads_rating</td>
  </tr>
  <tr>
    <td></td>
    <td>Goodreads Reviews</td>
    <td>booklore:goodreads_review_count</td>
  </tr>
  <tr>
    <td></td>
    <td>Hardcover Rating</td>
    <td>booklore:hardcover_rating</td>
  </tr>
  <tr>
    <td></td>
    <td>Hardcover Reviews</td>
    <td>booklore:hardcover_review_count</td>
  </tr>
  <tr>
    <td></td>
    <td>Lubimyczytac Rating</td>
    <td>booklore:lubimyczytac_rating</td>
  </tr>
  <tr>
    <td></td>
    <td>Ranobedb Rating</td>
    <td>booklore:ranobedb_rating</td>
  </tr>
</table>
<br>

## 🧪 Testing Strategy
Docker image for testing with Fix applied can be downloaded here:
``` code
docker pull jadehawk/jh-booklore:1.18.5-metadata-saver
```
Also added extra JUNIT test so we can test extracting all the custom columns using **"gradlew Test"**, all current test pass.

## 📸 Visual Changes _(if applicable)_
<!-- Attach screenshots or videos demonstrating UI/UX modifications -->


## ⚠️ Required Pre-Submission Checklist
<!-- ⛔ Pull requests will NOT be considered for review unless ALL required items are completed -->
<!-- All items below are MANDATORY prerequisites for submission -->
- [x] Code adheres to project style guidelines and conventions
- [x] Branch synchronized with latest `develop` branch
- [x] Automated unit/integration tests added/updated to cover changes
- [x] All tests pass locally (`./gradlew test` for backend)
- [x] Manual testing completed in local development environment
- [ ] Flyway migration versioning follows correct sequence _(if database schema modified)_
- [ ] Documentation pull request submitted to [booklore-docs](https://github.com/booklore-app/booklore-docs) _(required for features or enhancements that introduce user-facing or visual changes)_


## 💬 Additional Context _(optional)_
<!-- Provide any supplementary information, implementation considerations, or discussion points for reviewers -->

balazs-szucs commented on .dockerignore:
```diff
@@ -0,0 +1,34 @@
+# Dev config
+booklore-api/src/main/resources/application-local.yaml
+local/
+shared/
```
Unrelated change

balazs-szucs commented on booklore-api/src/main/java/com/adityachandel/booklore/service/metadata/extractor/EpubMetadataExtractor.java:
```diff
@@ -35,6 +54,103 @@ public class EpubMetadataExtractor implements FileMetadataExtractor {
+     * IMPORTANT: Calibre requires custom column lookup names to be LOWERCASE.
+     * This mapping includes both camelCase and lowercase versions for compatibility.
+     */
+    private static final Map<String, BiConsumer<BookMetadata.BookMetadataBuilder, String>> CALIBRE_FIELD_MAPPINGS = Map.ofEntries(
```
JUnit tests for this, perhaps?

jadehawk commented:
@balazs-szucs I've addressed all the review feedback:

- Eliminated DRY violations with helper methods for identifier extraction
- Fixed Build-to-Check anti-pattern using processedFields Set tracking
- Moved regex patterns to static constants to avoid recompilation
- Replaced magic numbers with calculated prefix lengths
- Consolidated repetitive Set field and pagecount parsing patterns
- Removed excessive documentation and comments
- Removed unrelated file changes (.dockerignore and .gitignore entries)

The refactoring reduces the code by 207 lines (129 insertions, 336 deletions) while maintaining all functionality. Code has been tested and works without issues. Ready for re-review.

MightyHalfGinger commented on booklore-api/src/main/java/com/adityachandel/booklore/service/metadata/extractor/EpubMetadataExtractor.java:
```diff
@@ -173,33 +269,36 @@ public BookMetadata extractMetadata(File epubFile) {
                             }
                             case "creator" -> authors.add(text);
-                            case "subject" -> categories.add(text);
+                            case "subject" -> tags.add(text);
```
This will cause the epub writer to be a bit confused since it saves the categories list to the subject tag: https://github.com/booklore-app/booklore/blob/36d5f9e6e059b5fd864642354fba5b21cfff1297/booklore-api/src/main/java/com/adityachandel/booklore/service/metadata/writer/EpubMetadataWriter.java#L100-L108

MightyHalfGinger commented on booklore-api/src/main/java/com/adityachandel/booklore/service/metadata/extractor/EpubMetadataExtractor.java:
```diff
@@ -379,4 +596,60 @@ private byte[] extractFileFromZip(File epubFile, String path) {
+                    }
+                    
+                    if ("#categories".equals(fieldName)) {
+                        extractSetField(value, categories);
```
Could we use `extractAndSetUserMetadataSet(value, builder::categories);` here rather than needing to pass in the categories set from the caller?

jadehawk commented:
@MightyHalfGinger 

Ok, made the following changes.
If user has created a custom column called "genres" or "categories" and has a VALUE in that column then we map that value to the "Genres" in booklore.

If user has created a custom column called "genres" or "categories" and has a NO VALUE in that column then we map the value of "tags" to the "Genres" which is the current booklore behavior.

If user does not have any #genres or #categories custom column, we map "tags" to "Genres" in booklore which is current booklore behavior.

Essentially, leaving all tags to genres the same, unless user creates their custom columns in Calibre.

MightyHalfGinger commented on booklore-api/src/main/java/com/adityachandel/booklore/service/metadata/extractor/EpubMetadataExtractor.java:
```diff
@@ -184,9 +287,17 @@ public BookMetadata extractMetadata(File epubFile) {
                             }
                             case "creator" -> {
-                                String role = el.getAttributeNS(OPF_NS, "role");
+                                String role = el.getAttributeNS("http://www.idpf.org/2007/opf", "role");
```
Lets keep the constant for the namespace

MightyHalfGinger commented on booklore-api/src/main/java/com/adityachandel/booklore/service/metadata/extractor/EpubMetadataExtractor.java:
```diff
@@ -362,8 +612,7 @@ private String findCoverImageHrefInOpf(File epubFile) {
                             String href = item.getAttribute("href");
-                            String decodedHref = URLDecoder.decode(href, StandardCharsets.UTF_8);
-                            return resolvePath(opfPath, decodedHref);
+                            return resolvePath(opfPath, href);
```
Why did this get changed?

MightyHalfGinger commented on booklore-api/src/test/java/com/adityachandel/booklore/service/metadata/writer/EpubMetadataWriterTest.java:
```diff
@@ -55,36 +52,53 @@ void setUp() {
-                    %s
-                </package>
-                """, existingMetadata);
-            File epubFile = createEpubWithOpf(opfContent, "test-metadata-" + System.nanoTime() + ".epub");
```
Another bad merge I think since this testcase seems to be removed

MightyHalfGinger commented on booklore-api/src/main/java/com/adityachandel/booklore/service/metadata/extractor/EpubMetadataExtractor.java:
```diff
@@ -21,22 +39,85 @@
+            }
+        }),
+        Map.entry("#series_name", (builder, value) -> builder.seriesName(value)),
+        Map.entry("#series_number", (builder, value) -> safeParseFloat(value, builder::seriesNumber)),
```
calibre already has well-defined fields for this rather than custom columns so I don't think we should add extra support for this.

MightyHalfGinger commented on booklore-api/src/main/java/com/adityachandel/booklore/service/metadata/extractor/EpubMetadataExtractor.java:
```diff
@@ -156,24 +238,18 @@ public BookMetadata extractMetadata(File epubFile) {
                                     }
                                 }
+                                
+                                if ("calibre:tags".equals(name)) {
```
Im not seeing this field ever used by calibre... Is there an example on how this can get set?

MightyHalfGinger commented on booklore-api/src/main/java/com/adityachandel/booklore/service/metadata/extractor/EpubMetadataExtractor.java:
```diff
@@ -410,4 +619,94 @@ private byte[] extractFileFromZip(File epubFile, String path) {
+                        extractAndSetUserMetadataSet(value, builder::authors);
+                        continue;
+                    }
+                    
```
Why would authors ever appear as an extra column when we have native support for it?

MightyHalfGinger commented on booklore-api/src/main/java/com/adityachandel/booklore/service/metadata/extractor/EpubMetadataExtractor.java:
```diff
@@ -198,31 +282,26 @@ public BookMetadata extractMetadata(File epubFile) {
                                 }
                             }
-                            case "subject" -> categories.add(text);
+                            case "subject" -> tags.add(text);
```
This should continue to be saved to categories as before

MightyHalfGinger commented on booklore-api/src/main/java/com/adityachandel/booklore/service/metadata/extractor/EpubMetadataExtractor.java:
```diff
@@ -21,22 +39,85 @@
+        }
+    }
+
+    private static final List<IdentifierMapping> IDENTIFIER_PREFIX_MAPPINGS = List.of(
```
Can we add the douban_id here rather than in the customer user_metadata?

MightyHalfGinger commented on booklore-api/src/main/java/com/adityachandel/booklore/service/metadata/extractor/EpubMetadataExtractor.java:
```diff
@@ -101,13 +182,14 @@ public BookMetadata extractMetadata(File epubFile) {
 
                     NodeList children = metadata.getChildNodes();
-
+                    
```
Why the whitespace change?

MightyHalfGinger commented on booklore-api/src/main/java/com/adityachandel/booklore/service/metadata/extractor/EpubMetadataExtractor.java:
```diff
@@ -140,9 +222,9 @@ public BookMetadata extractMetadata(File epubFile) {
                                     titleTypeById.put(refines.substring(1), content.toLowerCase());
                                 }
-
+                                
```
Why the whitespace change?

MightyHalfGinger commented on booklore-api/src/main/java/com/adityachandel/booklore/service/metadata/extractor/EpubMetadataExtractor.java:
```diff
@@ -1,5 +1,22 @@
+import java.sql.*;
+import java.util.*;
+import java.util.regex.Pattern;
+
```
Can we put these with the other imports rather than separately?

jadehawk commented:
> I finally sat down and went through this huge change in total, and I'm a bit concerned about the complexity/overreliance on user_metadata from calibre here. The whole point is they are custom for a "user", and not something more standard that epubs usually work with. There are certain usecases like the reviews and such where it makes a lot of sense to store the data there, but in many of the other cases we have existing epub standards that we should be using instead so that epub readers can hopefully make use of the data.
> 
> As an example, the more I think about this the less I think we should really need to add such complex support for the genres... Looking at how calibre works, it just saves the `tags` section directly to the subject in the opf file without any more thought. I think we should just follow along here and leave it so that calibre's "tags" are equivalent to booklore's "genres". I know they can be completely arbitrary, but having tried to fetch metadata for various books, the genre data from goodreads/hardcover/etc. can also be all over the place rather than strict "genres". This way booklore's separate tags can be saved to some other attribute in the epub rather than the subjects (e.g. something like `booklore:tags`).
> 
> As a suggestion:
> 
> * booklore's genre/categories =writes to=> epub's `subject` =read as=> calibre's tags
> * booklore's tags =writes to=> epub's `booklore:tags` and `calibre:user_metadata#extra_tags` =read as=> calibre custom extra_tags data
> 
> That way the data is preserved on both sides and should survive imports in both directions while fitting in with most epub readers that will only support subjects.
> 
> > You raised a great question here. I did not use that function because of the way I was generating the "categories" value, and when using the builder, it puts the first value found when called and ignores the rest.
> 
> The issue seems to stem from us effectively "double calling" the builder functions in both the outer logic where we get it from the `calibre:tags` and the `subject` metadata, and then again from the logic inside `extractCalibreUserMetadata`. I made a quick mock setup locally where I called builderMeta.authors multiple times with different data each time and only the last set of values stuck as you stated.
> 
> Looking at [the lombok docs](https://projectlombok.org/features/Builder) it seems this is the default behaviour for fields, but it can be changed for collections like we have here by adding a `@Singular` tag to the attribute in `BookMetadata.java`. That seems to fix my mock setup, so perhaps that will fix what you're referring to here as well, if we want to have them be combined.

@MightyHalfGinger

Happy New Year!

Thank you for reviewing the code and providing feedback on what needed to be changed or refactored.

I agree with all your points and have adjusted my approach. I moved away from the overly complex #genre / #categories extraction from Calibre and simplified things by continuing to treat dc:subject the same way Booklore currently does—mapping it directly to Booklore’s Genres (categories).

Our focus is now strictly on importing Calibre’s “Ratings & Reviews” (Amazon, Goodreads, Hardcover) when those custom columns exist, along with the corresponding IDs found in the Identifiers section. We’re also bringing in #extra_tags, #moods, and #subtitle. This covers all metadata Booklore currently tracks.

We’re still maintaining redundancy of Booklore metadata inside the EPUB using booklore:* fields. This ensures that if the database fails or a user recreates their Booklore library, all UI‑visible metadata remains preserved inside the EPUB. Once a book becomes part of Booklore’s library, all Calibre tags are removed and replaced with booklore:* tags, resulting in a much cleaner EPUB structure.

I also cleaned up whitespace, restored constants, and brought back code that was accidentally removed.

Regarding Douban metadata: I left that section out because DoubanBookParser.java already collects and maps all Douban metadata to the appropriate Booklore fields. However, doubanId, doubanRating, and doubanReviewCount are discarded during database entry since they don’t exist in BookMetadataEntity. This might be an incomplete implementation or intentional design—hard to say. To persist Douban metadata, I’d need to add those fields to the database and complete the implementation.

About @singular: I initially used it as you suggested and it worked well. But when MetadataChangeDetectorTest was introduced in PR #1985 (Part of Release 1.16.1), it failed on all three fields where I used @singular (authors, tags, moods). Removing @singular resolved the issue and the tests passed. I also manually tested adding authors, saving, adding another author, saving again, and all authors were correctly written to the EPUB—not just the last one. This was after removing the complicated genre/category logic.

Not sure when you or other devs will have time to review the PR with the New Year, but until then—

Happy New Year, everyone!

acx10 commented:
@balazs-szucs Thanks man! I’ll take your word for it.

jadehawk commented:
@balazs-szucs I know this was merged and supposed to have been added on v2.0, But a complete refactor was done on the EpubMetadataExtractor.java before v2.0 was released essentially removing all logic from this PR replaced by your logic that looks neat, yet none of the metadata that was previously successfully extracted from the epub (processed in Calibre) is ever reaching Booklore's UI. Not even the simplest things such as Identifiers (ASIN,MOBI-ASIN,GOODREADS,GOOGLE,HARDCOVER......) Non are extracted.

Technically nullifying this PR completely.

I see a json sidecar was implemented in the release, this sidecar does not extract the metadata from the epub completely either. it's limited to Page count and title series and description only?? No identifiers. Was this done deliverately or an oversight? I happily updated to v2.0.4 from my custom image, to take advantage of the new updates only to find out the last 200 ebooks uploaded from calibre to booklore, non-had metadata extracted.

included is a TEST ebook (Scrambled) used to TEST ALL FIELDS (Minus the new Age & Content rating.
[The Dark Cygnet Files - #04.00.zip](https://github.com/user-attachments/files/25620824/The.Dark.Cygnet.Files.-.04.00.zip)

balazs-szucs commented:
@jadehawk I dont believe my refactor did that, 100% oversight (from somewhere, probably not my PR). From what I can tell, my PR is not supposed to have adverse affect on Calibre things. But I can take a look remedy this, shorty.

Very sorry about that.

balazs-szucs commented:
I also did NOT modify JUnit testing methods, ergo, if or when it failed, we were oblivious to it since, the testing suite still passed, indicating no apparent problems. :cry: Sadly, this is kinda how Software goes, unless, there is explicit tests for it, it really hard to gurantee no regression. JUnit tests _has_ to fail explicitly in order for us to easily catch these. Sorry again!

balazs-szucs commented:
Also, many thanks for letting me know this, I made the PR for this, I would hope this will be merged/released soon, please, feel free to keep me in the loop or ping if anything else pops-up.

balazs-szucs commented:
Also forgot to say: 
<img width="3060" height="1204" alt="image" src="https://github.com/user-attachments/assets/90cad173-7469-460b-acac-1467c616f2c3" />

this how it parsed in your sample epub. I'm assuming this is good? If you can take quick look if it missed anything it'd nice. Thanks!
